feat(blog): Using Themes for Distributed Docs (#15330)
gatsbyjs/gatsby: docs/blog/2019-07-03-using-themes-for-distributed-docs/index.md

![Many sources going through one website](./images/many-to-one.png)

An alternative approach would be for each repo to deploy a piece of the website based on some shared configuration, allowing each of them to exist independently but remain up-to-date with changes to the website layout and infrastructure. In this world, changes to one repo’s documentation only means rebuilding and deploying its small part of the greater website. The goal of this approach is to reduce or remove the bottlenecks involved in making docs changes.

![Each source has its own website](./images/one-to-one.png)

Gatsby Themes are a great choice for this task. Themes allow us to build a full-featured Gatsby website—configuring plugins, writing custom components, using GraphQL queries—one time and apply it to multiple data sources. Themes are highly configurable, so each website can have its own defining features, like a page title, description, and more.

> In this blog post, I’m going to talk about some of our favourite parts about using Gatsby Themes to deliver a consistent docs experience.

Reviewer: ```suggestion
In this blog post, I’m going to talk about some of our favorite parts about using Gatsby Themes to deliver a consistent docs experience.
```
Updating per the [Gatsby Style Guide](https://www.gatsbyjs.org/contributing/gatsby-style-guide/#prefer-us-english).

## Versioning with `gatsby-source-git`

We were able to implement versioned docs by taking advantage of the way that theme configuration works. Normally, a website’s `gatsby-config.js` file would export a JavaScript object, but Themes can include a `gastby-config.js` that exports a function that takes options passed to it by the theme **consumer**, and returns a Gatsby configuration object. You can see how we use this pattern in [our theme’s `gatsby-config.js`](https://github.com/apollographql/gatsby-theme-apollo/blob/master/packages/gatsby-theme-apollo-docs/gatsby-config.js).

The Apollo docs theme accepts a mapping of version numbers to git branches, which it passes along to the `gatsby-source-git` plugin. A theme consumer can specify minimal configuration, and the theme takes care of creating pages and facilitating navigation between versions.

## Leveraging Gatsby Remark plugins

Gatsby’s preeminent Markdown parser, `gatsby-transformer-remark` features a rich ecosystem of useful plugins to do things like code highlighting, adding anchors to headings, and organizing images.

> Since Gatsby Remark plugins allow us to look at the Markdown AST of our pages at build time, we were able to create an internal link checker that scans our Markdown for anchor elements and compares them to known page slugs and heading IDs. This helps us to avoid deploying updates containing broken links, without the need to host the changes and check links with third-party software. You can see how we achieve this in [`gatsby-remark-check-links`](https://github.com/trevorblades/gatsby-remark-check-links).

Reviewer: We need this in the Gatsby docs! 😄

## Sharing components and more

> Using [MDX](https://gatsby-mdx.netlify.com/), we’re able to write rich documentation by including React components within a Markdown document. Since `gatsby-mdx` uses Remark under the hood, all of the Remark plugin features that were mentioned in the last section still apply.

Reviewer: ```suggestion
Using [MDX](/docs/mdx/), we’re able to write rich documentation by including React components within a Markdown document. Since `gatsby-plugin-mdx` uses Remark under the hood, all of the Remark plugin features that were mentioned in the last section still apply.
```
Updating to use the official Gatsby docs and the latest distributed plugin!

![Rendering components in MDX](./images/mdx-components.png)

> `gatsby-mdx` also allows us to replace Markdown elements with custom React components using the [`components` prop on the `MDXProvider` component](https://gatsby-mdx.netlify.com/api-reference/mdx-provider). We use this feature to enhance our code blocks with copy buttons, filenames, and multiple language options.

Reviewer: We should link to the Gatsby docs wherever possible. There is #14890 on the way, and it would be good to update this MDXProvider docs link when that migration work is done.
```suggestion
`gatsby-plugin-mdx` also allows us to replace Markdown elements with custom React components using the [`components` prop on the `MDXProvider` component](https://gatsby-mdx.netlify.com/api-reference/mdx-provider). We use this feature to enhance our code blocks with copy buttons, filenames, and multiple language options.
```
